File: packages/tonic-core/src/hydration/llmRefinement.ts
export type LlmChatParams = {
  baseUrl: string;
  model: string;
  apiKey: string;
  system: string;
  user: string;
  timeoutMs: number;
  /** When false, omit OpenAI `response_format` (some OpenAI-compatible servers reject json_object). */
  jsonObject?: boolean;
};

function stripJsonFence(s: string): string {
  const t = s.trim();
  const fence = /^```(?:json)?\s*([\s\S]*?)```$/m.exec(t);
  if (fence) {
    return fence[1]!.trim();
  }
  return t;
}

export async function callOpenAiCompatibleJson(
  params: LlmChatParams,
  fetchFn: typeof fetch = fetch,
): Promise<{ ok: true; text: string } | { ok: false; message: string }> {
  const url = `${params.baseUrl.replace(/\/$/, "")}/chat/completions`;
  const ac = new AbortController();
  const tid = setTimeout(() => ac.abort(), Math.max(1000, params.timeoutMs));
  const useJsonObject = params.jsonObject !== false;
  try {
    const body: Record<string, unknown> = {
      model: params.model,
      messages: [
        { role: "system", content: params.system },
        { role: "user", content: params.user },
      ],
      temperature: 0.2,
    };
    if (useJsonObject) {
      body.response_format = { type: "json_object" };
    }
    const res = await fetchFn(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${params.apiKey}`,
      },
      body: JSON.stringify(body),
      signal: ac.signal,
    });
    if (!res.ok) {
      const errText = await res.text();
      return { ok: false, message: `HTTP ${res.status}: ${errText.slice(0, 500)}` };
    }
    const data = (await res.json()) as {
      choices?: Array<{ message?: { content?: string } }>;
    };
    const text = data.choices?.[0]?.message?.content ?? "";
    if (!text) {
      return { ok: false, message: "empty completion" };
    }
    return { ok: true, text: stripJsonFence(text) };
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    return { ok: false, message: msg };
  } finally {
    clearTimeout(tid);
  }
}

export type QuestionRefinementPartial = {
  refined_left_intent?: string;
  refined_right_intent?: string;
  merge_goals?: string[];
  assumptions?: string[];
  subquestions?: Array<{ id: string; text: string; priority?: number }>;
};

export function parseQuestionRefinementJson(
  raw: string,
  mode: "improver" | "subquestions",
): { ok: true; value: QuestionRefinementPartial } | { ok: false; message: string } {
  let j: unknown;
  try {
    j = JSON.parse(raw);
  } catch {
    return { ok: false, message: "invalid JSON from model" };
  }
  if (!j || typeof j !== "object") {
    return { ok: false, message: "model JSON must be object" };
  }
  const o = j as Record<string, unknown>;
  if (mode === "improver") {
    const left = typeof o.refined_left_intent === "string" ? o.refined_left_intent : "";
    const right = typeof o.refined_right_intent === "string" ? o.refined_right_intent : "";
    const mergeGoals = Array.isArray(o.merge_goals)
      ? o.merge_goals.filter((x): x is string => typeof x === "string")
      : [];
    const assumptions = Array.isArray(o.assumptions)
      ? o.assumptions.filter((x): x is string => typeof x === "string")
      : [];
    if (!left || !right) {
      return { ok: false, message: "improver JSON missing refined_left_intent/refined_right_intent" };
    }
    return {
      ok: true,
      value: {
        refined_left_intent: left,
        refined_right_intent: right,
        merge_goals: mergeGoals,
        assumptions,
      },
    };
  }
  const sq = o.subquestions;
  if (!Array.isArray(sq)) {
    return { ok: false, message: "subquestions array required" };
  }
  const subquestions = sq
    .map((row) => {
      if (!row || typeof row !== "object") {
        return null;
      }
      const r = row as Record<string, unknown>;
      const id = typeof r.id === "string" ? r.id : "";
      const text = typeof r.text === "string" ? r.text : "";
      const priority = typeof r.priority === "number" ? r.priority : 0;
      if (!id || !text) {
        return null;
      }
      return { id, text, priority };
    })
    .filter((x): x is { id: string; text: string; priority: number } => x !== null)
    .sort((a, b) => (a.priority !== b.priority ? b.priority - a.priority : a.id.localeCompare(b.id)));
  return { ok: true, value: { subquestions } };
}

File: merge-tonic-lib/tonic/hydration/llm_refinement.py
"""OpenAI-compatible chat.completions (parity with llmRefinement.ts)."""

from __future__ import annotations

import json
import os
import re
import ssl
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Any, Literal


@dataclass
class LlmChatParams:
    base_url: str
    model: str
    api_key: str
    system: str
    user: str
    timeout_ms: int
    json_object: bool = True


def _strip_json_fence(s: str) -> str:
    t = s.strip()
    m = re.match(r"^```(?:json)?\s*([\s\S]*?)```$", t, re.MULTILINE)
    if m:
        return m.group(1).strip()
    return t


def call_openai_compatible_json(params: LlmChatParams) -> tuple[bool, str]:
    url = params.base_url.rstrip("/") + "/chat/completions"
    payload: dict[str, Any] = {
        "model": params.model,
        "messages": [
            {"role": "system", "content": params.system},
            {"role": "user", "content": params.user},
        ],
        "temperature": 0.2,
    }
    if params.json_object:
        payload["response_format"] = {"type": "json_object"}
    body = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=body,
        method="POST",
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {params.api_key}",
        },
    )
    try:
        ctx = ssl.create_default_context()
        with urllib.request.urlopen(req, timeout=max(1.0, params.timeout_ms / 1000.0), context=ctx) as resp:
            raw = resp.read().decode("utf-8", errors="replace")
    except urllib.error.HTTPError as e:
        err = e.read().decode("utf-8", errors="replace")[:500]
        return False, f"HTTP {e.code}: {err}"
    except Exception as e:
        return False, str(e)

    try:
        data: dict[str, Any] = json.loads(raw)
    except json.JSONDecodeError:
        return False, "invalid JSON from API"

    choices = data.get("choices")
    if not isinstance(choices, list) or not choices:
        return False, "empty completion"
    msg = choices[0]
    if not isinstance(msg, dict):
        return False, "empty completion"
    content = msg.get("message", {})
    text = content.get("content", "") if isinstance(content, dict) else ""
    if not isinstance(text, str) or not text.strip():
        return False, "empty completion"
    return True, _strip_json_fence(text)


def _hf_chat_content(resp: Any) -> str:
    if resp is None:
        return ""
    choices = getattr(resp, "choices", None)
    if choices is None and isinstance(resp, dict):
        choices = resp.get("choices")
    if not choices:
        return ""
    ch0 = choices[0]
    msg = getattr(ch0, "message", None)
    if msg is None and isinstance(ch0, dict):
        msg = ch0.get("message")
    if isinstance(msg, dict):
        c = msg.get("content")
        return c if isinstance(c, str) else ""
    if msg is not None:
        c = getattr(msg, "content", None)
        return c if isinstance(c, str) else ""
    return ""


def call_hf_inference_json(params: LlmChatParams) -> tuple[bool, str]:
    try:
        from huggingface_hub import InferenceClient
    except Exception as e:
        return False, f"huggingface_hub import failed: {e}"
    token = (os.environ.get("HF_TOKEN") or params.api_key or "").strip()
    if not token:
        return False, "HF_TOKEN (or api_key) required for hf_inference LLM"
    model = (os.environ.get("TONIC_HF_CHAT_MODEL") or params.model or "").strip()
    if not model:
        return False, "TONIC_HF_CHAT_MODEL or pipeline llm model required for hf_inference"
    provider = (os.environ.get("TONIC_HF_INFERENCE_PROVIDER") or "").strip() or None
    client = InferenceClient(api_key=token, provider=provider)
    messages = [
        {"role": "system", "content": params.system},
        {"role": "user", "content": params.user},
    ]
    if params.json_object:
        try:
            resp = client.chat_completion(
                model=model,
                messages=messages,
                max_tokens=2048,
                response_format={"type": "json_object"},
            )
            text = _hf_chat_content(resp)
            if text.strip():
                return True, _strip_json_fence(text)
        except Exception:
            pass
    try:
        resp = client.chat_completion(model=model, messages=messages, max_tokens=2048)
        text = _hf_chat_content(resp)
        if not text.strip():
            return False, "empty hf_inference completion"
        return True, _strip_json_fence(text)
    except Exception as e:
        return False, str(e)


def call_llm_json(params: LlmChatParams) -> tuple[bool, str]:
    backend = (os.environ.get("TONIC_LLM_BACKEND") or "").strip().lower()
    if backend in ("hf_inference", "hf-inference"):
        return call_hf_inference_json(params)
    return call_openai_compatible_json(params)


def parse_question_refinement_json(
    raw: str, mode: Literal["improver", "subquestions"]
) -> tuple[bool, str, dict[str, Any]]:
    try:
        j: Any = json.loads(raw)
    except json.JSONDecodeError:
        return False, "invalid JSON from model", {}
    if not isinstance(j, dict):
        return False, "model JSON must be object", {}
    if mode == "improver":
        left = j.get("refined_left_intent") if isinstance(j.get("refined_left_intent"), str) else ""
        right = j.get("refined_right_intent") if isinstance(j.get("refined_right_intent"), str) else ""
        mg = j.get("merge_goals")
        ass = j.get("assumptions")
        merge_goals = [x for x in mg if isinstance(x, str)] if isinstance(mg, list) else []
        assumptions = [x for x in ass if isinstance(x, str)] if isinstance(ass, list) else []
        if not left or not right:
            return False, "improver JSON missing refined_left_intent/refined_right_intent", {}
        return True, "", {
            "refined_left_intent": left,
            "refined_right_intent": right,
            "merge_goals": merge_goals,
            "assumptions": assumptions,
        }
    sq = j.get("subquestions")
    out_sq: list[dict[str, Any]] = []
    if isinstance(sq, list):
        for item in sq:
            if not isinstance(item, dict):
                continue
            sid = item.get("id")
            stext = item.get("text")
            if not isinstance(sid, str) or not isinstance(stext, str):
                continue
            pr = item.get("priority")
            entry: dict[str, Any] = {"id": sid, "text": stext}
            if isinstance(pr, (int, float)):
                entry["priority"] = int(pr)
            out_sq.append(entry)
    out_sq.sort(key=lambda x: (x.get("priority", 0), x.get("id", "")))
    if not out_sq:
        return False, "subquestions JSON missing non-empty subquestions", {}
    return True, "", {"subquestions": out_sq[:8]}
